File: frontend/lib/models/live_session.dart
/// Live Session Model
/// Represents a BigBlueButton live session scheduled by a teacher
class LiveSession {
  final String id;
  final String teacherId;
  final String courseId;
  final String sectionId;
  final String? lessonId;
  final String title;
  final String? description;
  final DateTime scheduledAt;
  final int duration;
  final DateTime? expectedEndTime;
  final String? bbbMeetingId;
  final String? bbbInternalMeetingId;
  final String? bbbModeratorUrl;
  final String? bbbAttendeeUrl;
  final String? bbbModeratorPw;
  final String? bbbAttendeePw;
  final String status;
  final String? recordingUrl;
  final int recordingDuration;
  final String? recordingFormat;
  final DateTime? endedAt;
  final int maxParticipants;
  final SessionSettings settings;
  final int participantCount;
  final List<dynamic>? attendees;
  final List<dynamic>? attendedAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  // Populated fields
  final Map<String, dynamic>? course;
  final Map<String, dynamic>? section;
  final Map<String, dynamic>? lesson;
  final Map<String, dynamic>? teacher;

  LiveSession({
    required this.id,
    required this.teacherId,
    required this.courseId,
    required this.sectionId,
    this.lessonId,
    required this.title,
    this.description,
    required this.scheduledAt,
    required this.duration,
    this.expectedEndTime,
    this.bbbMeetingId,
    this.bbbInternalMeetingId,
    this.bbbModeratorUrl,
    this.bbbAttendeeUrl,
    this.bbbModeratorPw,
    this.bbbAttendeePw,
    this.status = 'scheduled',
    this.recordingUrl,
    this.recordingDuration = 0,
    this.recordingFormat,
    this.endedAt,
    this.maxParticipants = 100,
    required this.settings,
    this.participantCount = 0,
    this.attendees,
    this.attendedAt,
    this.createdAt,
    this.updatedAt,
    this.course,
    this.section,
    this.lesson,
    this.teacher,
  });

  factory LiveSession.fromJson(Map<String, dynamic> json) {
    return LiveSession(
      id: json['_id'] ?? json['id'] ?? '',
      teacherId: (json['teacherId'] is Map
          ? json['teacherId']['_id'] ?? json['teacherId']['id'] ?? ''
          : json['teacherId'] ?? '') as String,
      courseId: (json['courseId'] is Map
          ? json['courseId']['_id'] ?? json['courseId']['id'] ?? ''
          : json['courseId'] ?? '') as String,
      sectionId: (json['sectionId'] is Map
          ? json['sectionId']['_id'] ?? json['sectionId']['id'] ?? ''
          : json['sectionId'] ?? '') as String,
      lessonId: json['lessonId'] is Map
          ? (json['lessonId']['_id'] ?? json['lessonId']['id']) as String?
          : json['lessonId'] as String?,
      title: json['title'] is String ? json['title'] : '',
      description: json['description'] is String ? json['description'] : null,
      scheduledAt: json['scheduledAt'] != null
          ? DateTime.parse(json['scheduledAt']).toLocal()
          : DateTime.now(),
      duration: json['duration'] ?? 60,
      expectedEndTime: json['expectedEndTime'] != null
          ? DateTime.parse(json['expectedEndTime']).toLocal()
          : null,
      bbbMeetingId: json['bbbMeetingId'] is String ? json['bbbMeetingId'] : null,
      bbbInternalMeetingId: json['bbbInternalMeetingId'] is String ? json['bbbInternalMeetingId'] : null,
      bbbModeratorUrl: json['bbbModeratorUrl'] is String ? json['bbbModeratorUrl'] : null,
      bbbAttendeeUrl: json['bbbAttendeeUrl'] is String ? json['bbbAttendeeUrl'] : null,
      bbbModeratorPw: json['bbbModeratorPw'] is String ? json['bbbModeratorPw'] : null,
      bbbAttendeePw: json['bbbAttendeePw'] is String ? json['bbbAttendeePw'] : null,
      status: json['status'] is String ? json['status'] : 'scheduled',
      recordingUrl: json['recordingUrl'] is String ? json['recordingUrl'] : null,
      recordingDuration: json['recordingDuration'] ?? 0,
      recordingFormat: json['recordingFormat'] is String ? json['recordingFormat'] : null,
      endedAt: json['endedAt'] != null
          ? DateTime.parse(json['endedAt']).toLocal()
          : null,
      maxParticipants: json['maxParticipants'] ?? 100,
      settings: json['settings'] != null
          ? SessionSettings.fromJson(json['settings'])
          : SessionSettings(),
      participantCount: json['participantCount'] ?? 0,
      attendees: json['attendees'] != null ? List<dynamic>.from(json['attendees']) : null,
      attendedAt: json['attendedAt'] != null ? List<dynamic>.from(json['attendedAt']) : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt']).toLocal()
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt']).toLocal()
          : null,
      course: json['courseId'] is Map ? json['courseId'] : null,
      section: json['sectionId'] is Map ? json['sectionId'] : null,
      lesson: json['lessonId'] is Map ? json['lessonId'] : null,
      teacher: json['teacherId'] is Map ? json['teacherId'] : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'teacherId': teacherId,
      'courseId': courseId,
      'sectionId': sectionId,
      'lessonId': lessonId,
      'title': title,
      'description': description,
      'scheduledAt': scheduledAt.toIso8601String(),
      'duration': duration,
      'bbbMeetingId': bbbMeetingId,
      'bbbInternalMeetingId': bbbInternalMeetingId,
      'bbbModeratorUrl': bbbModeratorUrl,
      'bbbAttendeeUrl': bbbAttendeeUrl,
      'bbbModeratorPw': bbbModeratorPw,
      'bbbAttendeePw': bbbAttendeePw,
      'status': status,
      'recordingUrl': recordingUrl,
      'recordingDuration': recordingDuration,
      'recordingFormat': recordingFormat,
      'endedAt': endedAt?.toIso8601String(),
      'maxParticipants': maxParticipants,
      'settings': settings.toJson(),
      'participantCount': participantCount,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  bool get isScheduled => status == 'scheduled';
  bool get isLive => status == 'live';
  bool get isEnded => status == 'ended';
  bool get isCancelled => status == 'cancelled';
  bool get hasRecording => recordingUrl != null && recordingUrl!.isNotEmpty;

  // Use backend-provided expectedEndTime if available, otherwise calculate it
  DateTime get calculatedEndTime => expectedEndTime ?? scheduledAt.add(Duration(minutes: duration));

  /// e.g. "Started 12 min ago · Ends at 15:30"
  /// or "Starts at 14:00 · Ends at 15:00"
  String get timeProgressInfo {
    final now = DateTime.now();
    final endTime = calculatedEndTime;
    final endStr = _hhmm(endTime);
    if (isLive || scheduledAt.isBefore(now)) {
      final elapsed = now.difference(scheduledAt);
      final elapsedStr = elapsed.inMinutes < 60
          ? '${elapsed.inMinutes} min ago'
          : '${elapsed.inHours}h ${elapsed.inMinutes % 60}m ago';
      return 'Started $elapsedStr · Ends at $endStr';
    } else {
      return 'Starts at ${_hhmm(scheduledAt)} · Ends at $endStr';
    }
  }

  static String _hhmm(DateTime dt) =>
      '${dt.hour.toString().padLeft(2, '0')}:${dt.minute.toString().padLeft(2, '0')}';

  String get formattedDuration {
    final hours = duration ~/ 60;
    final minutes = duration % 60;
    if (hours > 0) {
      return minutes > 0 ? '${hours}h ${minutes}m' : '${hours}h';
    }
    return '${minutes}m';
  }

  String get formattedScheduledDate {
    return '${scheduledAt.day}/${scheduledAt.month}/${scheduledAt.year}';
  }

  String get formattedScheduledTime {
    final now = DateTime.now();
    final difference = scheduledAt.difference(now);
    
    if (difference.inDays > 0) {
      return 'In ${difference.inDays} day${difference.inDays > 1 ? 's' : ''}';
    } else if (difference.inHours > 0) {
      return 'In ${difference.inHours} hour${difference.inHours > 1 ? 's' : ''}';
    } else if (difference.inMinutes > 0) {
      return 'In ${difference.inMinutes} min';
    } else if (difference.isNegative) {
      return 'Started';
    }
    return 'Starting soon';
  }
}

/// Session Settings Model
class SessionSettings {
  final bool enableChat;
  final bool enableWebcam;
  final bool muteOnEntry;
  final bool allowRecording;
  final bool waitingRoom;

  SessionSettings({
    this.enableChat = true,
    this.enableWebcam = true,
    this.muteOnEntry = true,
    this.allowRecording = true,
    this.waitingRoom = false,
  });

  factory SessionSettings.fromJson(Map<String, dynamic> json) {
    return SessionSettings(
      enableChat: json['enableChat'] ?? true,
      enableWebcam: json['enableWebcam'] ?? true,
      muteOnEntry: json['muteOnEntry'] ?? true,
      allowRecording: json['allowRecording'] ?? true,
      waitingRoom: json['waitingRoom'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'enableChat': enableChat,
      'enableWebcam': enableWebcam,
      'muteOnEntry': muteOnEntry,
      'allowRecording': allowRecording,
      'waitingRoom': waitingRoom,
    };
  }
}

File: backend/src/models/LiveSession.js
const mongoose = require('mongoose');

const liveSessionSchema = new mongoose.Schema({
  teacherId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Teacher ID is required']
  },
  courseId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Course',
    required: [true, 'Course ID is required']
  },
  sectionId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Section',
    required: [true, 'Section ID is required']
  },
  lessonId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Lesson',
    default: null
  },
  title: {
    type: String,
    required: [true, 'Session title is required'],
    trim: true,
    maxlength: [200, 'Title cannot exceed 200 characters']
  },
  description: {
    type: String,
    maxlength: [1000, 'Description cannot exceed 1000 characters']
  },
  scheduledAt: {
    type: Date,
    required: [true, 'Schedule date is required']
  },
  duration: {
    type: Number,
    required: [true, 'Duration is required'],
    min: [15, 'Duration must be at least 15 minutes'],
    max: [180, 'Duration cannot exceed 180 minutes']
  },
  expectedEndTime: {
    type: Date,
    default: null
  },
  bbbMeetingId: {
    type: String,
    default: null
  },
  bbbInternalMeetingId: {
    type: String,
    default: null
  },
  bbbModeratorUrl: {
    type: String,
    default: null
  },
  bbbAttendeeUrl: {
    type: String,
    default: null
  },
  bbbModeratorPw: {
    type: String,
    default: null
  },
  bbbAttendeePw: {
    type: String,
    default: null
  },
  status: {
    type: String,
    enum: ['scheduled', 'live', 'ended', 'cancelled'],
    default: 'scheduled'
  },
  recordingUrl: {
    type: String,
    default: null
  },
  recordingDuration: {
    type: Number,
    default: 0
  },
  recordingFormat: {
    type: String,
    enum: ['video', 'presentation', null],
    default: null
  },
  /** Every playback format BBB produced for this session. */
  recordingFormats: [{
    type: { type: String },
    url: String,
    duration: Number,
    size: Number
  }],
  /** Direct mp4, present only when the BBB server produces a video format. */
  recordingDownloadUrl: {
    type: String,
    default: null
  },
  recordingParticipants: {
    type: Number,
    default: 0
  },
  /**
   * The teacher reviews the recording, then decides whether students may keep
   * a copy. Watching is allowed as soon as it exists; downloading is opt-in,
   * matching how peer study sessions behave.
   */
  allowRecordingDownload: {
    type: Boolean,
    default: false
  },
  /** Lets a teacher hide a recording that did not come out usable. */
  isRecordingPublished: {
    type: Boolean,
    default: true
  },
  endedAt: {
    type: Date,
    default: null
  },
  maxParticipants: {
    type: Number,
    default: 100,
    min: [1, 'Must allow at least 1 participant'],
    max: [500, 'Cannot exceed 500 participants']
  },
  settings: {
    enableChat: {
      type: Boolean,
      default: true
    },
    enableWebcam: {
      type: Boolean,
      default: true
    },
    muteOnEntry: {
      type: Boolean,
      default: true
    },
    allowRecording: {
      type: Boolean,
      default: true
    },
    waitingRoom: {
      type: Boolean,
      default: false
    }
  },
  participantCount: {
    type: Number,
    default: 0
  },
  attendees: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  attendedAt: [{
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    joinedAt: {
      type: Date,
      default: Date.now
    },
    duration: {
      type: Number,
      default: 0 // in minutes
    }
  }],
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Indexes for better query performance
liveSessionSchema.index({ teacherId: 1, status: 1 });
liveSessionSchema.index({ courseId: 1, status: 1 });
liveSessionSchema.index({ sectionId: 1, lessonId: 1 });
liveSessionSchema.index({ scheduledAt: 1 });
liveSessionSchema.index({ status: 1, scheduledAt: 1 });
liveSessionSchema.index({ bbbMeetingId: 1 });

module.exports = mongoose.model('LiveSession', liveSessionSchema);
